File: sdks/official/fraiseql-php/src/Attributes/GraphQLType.php
<?php

declare(strict_types=1);

namespace FraiseQL\Attributes;

use Attribute;

/**
 * PHP 8 Attribute for defining GraphQL type definitions.
 *
 * Usage:
 * ```php
 * #[GraphQLType(name: 'User')]
 * class User {
 *     #[GraphQLField]
 *     public int $id;
 *
 *     #[GraphQLField]
 *     public string $name;
 * }
 * ```
 */
#[Attribute(Attribute::TARGET_CLASS)]
final readonly class GraphQLType
{
    /**
     * @param string|null $name Optional custom GraphQL type name. Defaults to class name.
     * @param string|null $sqlSource The SQL view backing this type (e.g. 'v_user').
     * @param string|null $description Optional description for schema documentation.
     * @param bool $isInput Whether this type represents a GraphQL input type.
     * @param bool $relay Whether this type implements the Relay Node interface.
     * @param bool $isError Whether this type represents a mutation error type.
     * @param bool $crud When true, auto-generate CRUD queries and mutations for this type.
     * @param bool $cascade When true, generated CRUD mutations include cascade support.
     * @param list<\FraiseQL\Relationship> $relationships Relationships to other types,
     *        followed by REST resource embedding (#1266) — `?select=orders(id,total)`,
     *        `?select=orders.count`, `?orders.status=paid` — and published in the served
     *        OpenAPI document and the generated client's `relationships` module.
     *
     * @throws \FraiseQL\FraiseQLException If a relationship name is declared twice; an
     *         embed resolves the first and the rest are unreachable, which no compiler
     *         diagnostic can attribute back to this declaration.
     */
    public function __construct(
        public ?string $name = null,
        public ?string $sqlSource = null,
        public ?string $description = null,
        public bool $isInput = false,
        public bool $relay = false,
        public bool $isError = false,
        public bool $crud = false,
        public bool $cascade = false,
        public array $relationships = [],
    ) {
        $seen = [];
        foreach ($relationships as $relationship) {
            if (!$relationship instanceof \FraiseQL\Relationship) {
                throw new \FraiseQL\FraiseQLException(
                    'relationships must be a list of FraiseQL\Relationship',
                );
            }
            if (isset($seen[$relationship->name])) {
                throw new \FraiseQL\FraiseQLException(sprintf(
                    'relationship "%s" is declared more than once; an embed resolves the '
                        . 'first and the rest are unreachable',
                    $relationship->name,
                ));
            }
            $seen[$relationship->name] = true;
        }
    }
}

File: sdks/official/fraiseql-java/src/main/java/com/fraiseql/core/GraphQLType.java
package com.fraiseql.core;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Marks a class as a GraphQL type for FraiseQL schema authoring.
 *
 * <p>Usage:
 * <pre>
 * &#64;GraphQLType
 * public class User {
 *     &#64;GraphQLField
 *     public int id;
 *
 *     &#64;GraphQLField
 *     public String name;
 * }
 * </pre>
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface GraphQLType {
    /**
     * Optional custom name for the GraphQL type.
     * If not specified, the class name is used.
     */
    String name() default "";

    /**
     * Optional description for the GraphQL type.
     */
    String description() default "";

    /**
     * Whether this type implements the Relay Node interface.
     * When true, the type participates in global object identification via node(id: ID!)
     * and can be used as the return type of relay-enabled list queries.
     * Requires pk_{entity} (BIGINT) to be present in the view's data JSONB.
     */
    boolean relay() default false;

    /**
     * The SQL view backing this type (e.g. "v_user").
     * Defaults to "v_" + snake_case(class name).
     */
    String sqlSource() default "";

    /**
     * When true, auto-generate CRUD queries and mutations for this type.
     */
    boolean crud() default false;

    /**
     * When true, generated CRUD mutations include cascade support.
     */
    boolean cascade() default false;
}
